### ISSUE TITLE
Capstone through Java binding triggers Invalid memory access error on Java 9 and above

### ISSUE THREAD

**Issue by mquigley** [aquynh/capstone#1246]
Using the test Java bindings project, Java 8 can successfully invoke Capstone. However, newer version of Java fail with Invalid memory access errors.

```
E:\dev\capstone_test>"C:\Program Files\Java\jdk1.8.0_72\bin\java.exe" -classpath "E:\dev\capstone_test\out\production\capstone_test;E:\dev\capstone_test\lib\jna-4.5.2.jar" Test
****************
Platform: X86 16bit (Intel syntax)
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00
Disasm:
0x1000:         lea     cx, word ptr [si + 0x32]
0x1003:         or      byte ptr [bx + di], al
... 
```

However, both Java 9 and 10 fail

```
E:\dev\capstone_test>"C:\Program Files\Java\jdk-9.0.1\bin\java.exe" -classpath E:\dev\capstone_test\out\production\capstone_test;E:\dev\capstone_test\lib\jna-4.5.2.jar Test
****************
Platform: X86 16bit (Intel syntax)
Code: 0x8d 0x4c 0x32 0x08 0x01 0xd8 0x81 0xc6 0x34 0x12 0x00 0x00
Disasm:
Exception in thread "main" java.lang.Error: Invalid memory access
        at com.sun.jna.Native.invokeInt(Native Method)
        at com.sun.jna.Function.invoke(Function.java:419)
        at com.sun.jna.Function.invoke(Function.java:354)
        at com.sun.jna.Library$Handler.invoke(Library.java:244)
        at capstone.$Proxy0.cs_disasm(Unknown Source)
        at capstone.Capstone.disasm(Capstone.java:458)
        at capstone.Capstone.disasm(Capstone.java:443)
        at Test.main(Test.java:167)
```
(also `"C:\Program Files\Java\jdk-10.0.2\bin\java.exe" -classpath E:\dev\capstone_test\out\production\capstone_test;E:\dev\capstone_test\lib\jna-4.5.2.jar Test`)

I've anecdotally heard that invalid memory access errors have gotten worse with newer JDKs, as newer Java versions have gotten more aggressive in memory use/reclaim.

**Comment by KYHSGeekCode** [comment]
On android I added cs_init_mem in the library that calls cs_option(mem) and added call to this function in Capston class's constructor.

**Comment by aquynh** [comment]
if you have a fix for this issue, please send pull req, thanks.

**Comment by BullyWiiPlaza** [comment]
[Java 11 also fails](https://stackoverflow.com/questions/53523121). +1 for a fix for this issue. :)

**Comment by mquigley** [comment]
It seems the issue is that the JNA library is describing `Native.LONG_SIZE == 4` and `Native.POINTER_SIZE == 8`.  The capstone C library is 64-bit. In `interface CS extends Library` the method to open the library and get the handle is `int cs_open(int arch, int mode, NativeLongByReference handle);`. This refers to `cs.c` `cs_err CAPSTONE_API cs_open(cs_arch arch, cs_mode mode, csh *handle)`. Inside that method, the `handle` is allocated with `ud = cs_mem_calloc(1, sizeof(*ud));`. 

This is where things get interesting. I put some printfs and verified that `sizeof(*ud)` is 8. On a <= 8 JVM, the returned pointer fits within 32-bits. On >= 9 JVM, the returned pointer is larger (e.g. Java 8 the handle pointer is `0000000001103510` while on 9 it's `000001EF8BE6F210`. I guess that earlier memory management schemes would create raw pointers closest to 0 in their own address space within the process, meaning you'd never go over 32-bits until the process allocated over 4GB. And on later JVMs the memory management got more creative for whatever purposes (like security) and the allocated pointers are obfuscated, or the native heap starts at 32-bit offset... I don't know, but it's certain that  `NativeLong` and `NativeLongByReference` for pointers is the problem here.

**TL;DR** - `NativeLong` and `NativeLongByReference` are used to store the handle which is an allocated pointer of size 8, not an actual 4 byte `long`, and the JNA library is truncating the handle pointer from 64-bits to 32-bits.

I might investigate more over the holiday weekend.

**Comment by mquigley** [comment]
Well that was easy! I changed the JNA bindings from `NativeLong` and `NativeLongByReference` to `NativePointer` and `NativePointerByReference` and everything seems to be working. I'll see about a pull request later.

**Comment by aquynh** [comment]
Cool, pls pull req

**Comment by mquigley** [comment]
Fixed with PR #1516

**Comment by PAX523** [comment]
@mquigley please take a look at https://github.com/aquynh/capstone/issues/1658

The binding's major version change makes it incompatible to be executed with the natives from 4.0.2.
